icu4c/source/common/mlbe.cpp

```diff
@@ -0,0 +1,455 @@
+    const int32_t MAX_FEATURE_LENGTH = 14;
+
+    bool isValid(const Element& element) {
+        return element.getLength() != 1 || element.getUblock()[0] != INVALID;
+    }
+
+    void concatChar(const char16_t *str, const UChar32 *arr, int32_t length, char16_t *feature, UErrorCode &status) {
+        if (U_FAILURE(status)) {
+            return;
+        }
+        UnicodeString result(str);
+        for (int i = 0; i < length; i++) {
+            result.append(arr[i]);
+        }
+        U_ASSERT(result.length() < MAX_FEATURE_LENGTH);
+        result.extract(feature, MAX_FEATURE_LENGTH, status);  // NUL-terminates
+    }
+
+    void writeString(const UnicodeString &str, char16_t *feature, UErrorCode &status) {
+        U_ASSERT(str.length() < MAX_FEATURE_LENGTH);
+        str.extract(feature, MAX_FEATURE_LENGTH, status);  // NUL-terminates
+    }
+}
+
+int32_t MlBreakEngine::divideUpRange(UText *inText, int32_t rangeStart, int32_t rangeEnd,
+                                       UVector32 &foundBreaks, const UnicodeString &inString,
+                                       const LocalPointer<UVector32> &inputMap,
+                                       UErrorCode &status) const {
+    if (U_FAILURE(status)) {
+        return 0;
+    }
+    if (rangeStart >= rangeEnd) {
+        status = U_ILLEGAL_ARGUMENT_ERROR;
+        return 0;
+    }
+
+    UVector32 boundary(inString.countChar32() + 1, status);
+    if (U_FAILURE(status)) {
+        return 0;
+    }
+    int32_t numBreaks = 0;
+    UChar32 ch;
+    UnicodeString index;
+    // The ML model groups six char to evaluate if the 4th char is a breakpoint.
+    // Like a sliding window, the elementList removes the first char and appends the new char from
+    // inString in each iteration so that its size always remains at six.
+    Element elementList[6];
+    if (U_FAILURE(status)) {
+        return 0;
+    }
+
+    int32_t codeUts = initElementList(inString, elementList, status);
+    int32_t length = inString.countChar32();
+    if (U_FAILURE(status)) return 0;
+    for (int32_t i = 1; i < length && U_SUCCESS(status); i++) {
+        evaluateBreakpoint(elementList, i, numBreaks, boundary, status);
+        if (i + 1 >= inString.countChar32()) break;
+        // Remove the first element and append a new element
+        uprv_memmove(elementList, elementList + 1, 5 * sizeof(Element));
+        ch = inString.countChar32(0, codeUts) < length ? inString.char32At(codeUts) : INVALID;
+        index = (ch != INVALID) ? getUnicodeBlock(ch, status) : UnicodeString(INVALID);
+        elementList[5].setCharAndUblock(ch, index);
+        if (ch != INVALID) {
+            codeUts += U16_LENGTH(ch);
+        }
+    }
+    if (U_FAILURE(status)) return 0;
+
+    // Add a break for the start if there is not one there already.
+    if (foundBreaks.size() == 0 || foundBreaks.peeki() < rangeStart) {
+        boundary.insertElementAt(0, 0, status);
+        numBreaks++;
+    }
```

Comment:
1. rangeStart is the index in the UText. Given a sentence with Japanese, digit and punctuation, only grouped Japanese would enter this engine so that rangeStart is not always zero.
For example,  UText="乗車率９０％程度だろうか", divideUpRange() will execute twice
1st: inString="乗車率",  (ragneStart, rangeEnd) = (0,3)
2nd: inString="程度だろうか",  (ragneStart, rangeEnd) = (6,12)
The test case in rbbitst.txt already contains punctuation and digits.

2. Actually, this implementation references CjkBreakEngine. Not very clear about the purpose, my understanding is to add 0 at the front.
https://github.com/unicode-org/icu/blob/main/icu4c/source/common/dictbe.cpp#L1393
I have changed to add 0 before the loop.